pkg/upgrade/upgrade.go

```diff
@@ -79,138 +90,292 @@ func NewUpgrade(c Config, l *zap.SugaredLogger) (*Upgrade, error) {
 
 // Run runs the operators installation process.
 func (u *Upgrade) Run(ctx context.Context) error {
-	if err := u.runEverestWizard(ctx); err != nil {
-		return err
+	// Get Everest version.
+	everestVersion, err := cliVersion.EverestVersionFromDeployment(ctx, u.kubeClient)
+	if err != nil {
+		return errors.Join(err, errors.New("could not retrieve Everest version"))
 	}
-	l, err := install.ValidateNamespaces(u.config.Namespaces)
+
+	// Check prerequisites
+	upgradeEverestTo, recVer, err := u.canUpgrade(ctx, everestVersion)
 	if err != nil {
+		if errors.Is(err, ErrNoUpdateAvailable) {
+			u.l.Info("You're running the latest version of Everest")
+			return nil
+		}
 		return err
 	}
-	u.config.NamespacesList = l
-	if err := u.upgradeOLM(ctx); err != nil {
+
+	// Start upgrade.
+	if err := u.upgradeOLM(ctx, recVer.OLM); err != nil {
 		return err
 	}
-	u.l.Info("Upgrading Percona Catalog")
-	if err := u.kubeClient.InstallPerconaCatalog(ctx); err != nil {
+
+	// We cannot use the latest version of catalog yet since
+	// at the time of writing, each catalog version supports only one Everest version.
+	catalogVersion := recVer.Catalog
+	if catalogVersion == nil {
+		u.l.Debugf("Percona catalog version was nil. Changing to %s", upgradeEverestTo)
+		catalogVersion = upgradeEverestTo
+	}
+	u.l.Infof("Upgrading Percona Catalog to %s", recVer.Catalog)
+	if err := u.kubeClient.InstallPerconaCatalog(ctx, catalogVersion); err != nil {
 		return err
 	}
-	u.l.Info("Percona Catalog has been upgraded")
-	u.l.Info("Patching subscriptions")
-	if err := u.patchSubscriptions(ctx); err != nil {
+
+	// Locate the correct install plan.
+	ctxTimeout, cancel := context.WithTimeout(ctx, 5*time.Minute)
+	defer cancel()
+
+	u.l.Info("Waiting for install plan for Everest operator")
+	ip, err := u.kubeClient.WaitForInstallPlan(
+		ctxTimeout, common.SystemNamespace,
+		common.EverestOperatorName, upgradeEverestTo,
+	)
+	if err != nil {
+		return errors.Join(err, errors.New("could not find install plan"))
+	}
+
+	u.l.Infof("Upgrading Everest to %s in namespace %s", upgradeEverestTo, common.SystemNamespace)
+	if err := u.kubeClient.InstallEverest(ctx, common.SystemNamespace, upgradeEverestTo); err != nil {
 		return err
 	}
```

Comment:
The logic for upgrade has been defined as:
1. Upgrade Everest
2. Upgrade Everest operator

In case we switch the order we will be in a situation where we have Everest operator running version `n+1` and Everest running version `n`. During this time Everest may not be compatible with the Everest operator which can cause issues.
In cases where the upgrade fails during this step, Everest may not be usable at all.

Upgrading Everest first helps us avoid this type of issue because Everest version is `n+1` and Everest operator is `n`. In this situation we can ensure Everest is compatible with Everest operator version `n+1` and `n`.

Why do you think we shall do this in a different order?